query/src/main/java/com/connexta/replication/query/QueryManager.java

```diff
@@ -0,0 +1,199 @@
+        serviceRefreshPeriod,
+        Executors.newSingleThreadScheduledExecutor());
+  }
+
+  QueryManager(
+      Stream<String> sites,
+      QueryServiceTools queryServiceTools,
+      long serviceRefreshPeriod,
+      ScheduledExecutorService executor) {
+    this.sites = sites.collect(Collectors.toSet());
+    this.queryServiceTools = queryServiceTools;
+    this.serviceRefreshPeriod = serviceRefreshPeriod;
+    this.executor = executor;
+  }
+
+  /**
+   * Initializes the query manager and kick start all query services to start polling sites that
+   * must be polled.
+   */
+  public void init() {
+    if (sites.isEmpty()) {
+      LOGGER.info("Managing queries for all configured sites.");
+    } else {
+      LOGGER.info("Managing queries for sites: {}.", sites);
+    }
+    executor.scheduleAtFixedRate(
+        this::reloadSiteConfigs, 0, serviceRefreshPeriod, TimeUnit.SECONDS);
+  }
+
+  /** Destroys the query manager and stop all currently running query services. */
+  public void destroy() {
+    LOGGER.info("Shutting down query managers");
+    executor.shutdownNow();
+    synchronized (this) {
+      for (QueryService service : services.values()) {
+        service.stop();
+      }
+      services.clear();
+    }
+  }
+
+  @VisibleForTesting
+  void reloadSiteConfigs() {
+    Stream<Site> siteStream = queryServiceTools.allSites();
+    synchronized (this) {
+      if (Thread.currentThread().isInterrupted()) {
+        return;
+      }
+
+      if (!sites.isEmpty()) {
+        siteStream = siteStream.filter(site -> sites.contains(site.getId()));
+      }
+      final Map<String, Site> newSites =
+          siteStream.collect(Collectors.toMap(Site::getId, Function.identity()));
+
+      // stop query services for sites that are no longer configured or have changed to a non-polled
+      // type
+      // update them if they are already running
+      // TODO: remove filters for regional DDF when we support other kinds of sites
+      updateQueryServices(newSites);
+
+      // start query services for new sites that must be polled
+      newSites.values().stream()
+          .filter(QueryManager::mustBePolled)
+          .filter(site -> site.getKind() == SiteKind.REGIONAL)
+          .map(this::newService)
+          .filter(Objects::nonNull)
+          .forEach(QueryService::start);
+    }
+  }
+
+  @VisibleForTesting
+  void updateQueryServices(Map<String, Site> newSites) {
+    final Iterator<Entry<String, QueryService>> serviceIterator = services.entrySet().iterator();
+    while (serviceIterator.hasNext()) {
+      final Entry<String, QueryService> entry = serviceIterator.next();
+      final Site updatedSite = newSites.remove(entry.getKey());
+
+      if (updatedSite != null
+          && updatedSite.getType().mustBePolled()
+          && updatedSite.getKind() == SiteKind.REGIONAL) {
+        try {
+          entry.getValue().update(updatedSite);
+        } catch (AdapterException | IllegalArgumentException e) {
```

Comment:
All we're doing is creating an adapter here. Can that throw an AdapterInterruptedException?

Reply:
It might be safe to still handle them wherever you handle `AdapterException`. I would rather not rely on how the code of that other class is but more on the fact that it documents it can throw that base AdapterException which in the end, could be any subclasses.